### ISSUE TITLE
Stub process is unhealthy and it will be restarted

### ISSUE THREAD

**Issue by swapkh91** [triton-inference-server/server#3678]
I'm getting `Stub process is unhealthy and it will be restarted` repeatedly when calling `infer`, after which the server restarts. I have deployed triton server on GKE with 3 models.

1st time when I infer `model1` I get this error, 2nd and consequent hits don't give this error. But if I infer `model2` after getting successful result from `model1` then again this error pops up and so on for `model3`.

logs:
```
responses.append(self.triton_client.infer(
      File "/home/swapnesh/triton/triton_env/lib/python3.8/site-packages/tritonclient/grpc/__init__.py", line 1086, in infer
        raise_error_grpc(rpc_error)
      File "/home/swapnesh/triton/triton_env/lib/python3.8/site-packages/tritonclient/grpc/__init__.py", line 61, in raise_error_grpc
        raise get_error_grpc(rpc_error) from None
    tritonclient.utils.InferenceServerException: [StatusCode.INTERNAL] Failed to process the request(s) for model instance 'damage_0', message: Stub process is not healthy.
```

I'm loading 3 models, using the python backend and custom triton image (converted detectron models) which I've built using this Dockerfile:
```
FROM nvcr.io/nvidia/tritonserver:21.10-py3

RUN pip3 install torch==1.9.1 torchvision==0.10.1 torchaudio==0.9.1 && \
    pip3 install pillow
```

Also, while running triton server locally using docker, I had to increase the `shm-size` as it was giving error to increase it from 64MB. On Kubernetes its a little tricky, to use `emptyDir` with `Memory` medium. My yaml looks like this:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: triton-mms
  name: triton-mms
spec:
  replicas: 1
  selector:
    matchLabels:
      app: triton-mms
  template:
    metadata:
      labels:
        app: triton-mms
    spec:
      containers:
      - image: <custom triton image>
        command: ["/bin/sh", "-c"]
        args: ["tritonserver --model-repository=<gcs model repo>"]
        imagePullPolicy: IfNotPresent
        name: triton-mms
        ports:
        - containerPort: 8000
          name: http-triton
        - containerPort: 8001
          name: grpc-triton
        - containerPort: 8002
          name: metrics-triton
        env:
        - name: GOOGLE_APPLICATION_CREDENTIALS
          value: /secret/gcp-creds.json
        resources:
          limits:
            memory: 5Gi
            nvidia.com/gpu: 1
          requests:
            memory: 5Gi
            nvidia.com/gpu: 1
        volumeMounts:
        - mountPath: /dev/shm
          name: dshm
        - name: vsecret
          mountPath: "/secret"
          readOnly: true
      volumes:
      - name: dshm
        emptyDir:
          medium: Memory
          sizeLimit: "1024Mi"
      - name: vsecret
        secret:
          secretName: gcpcreds
```

Never faced this issue before and I'm thinking it might be related to `shared memory` as I've never seen that error too.

**Comment by Tabrizian** [comment]
@swapkh91 Can you try 21.11 version of Triton? Also, could you share your model.py so that we can take a closer look?

**Comment by swapkh91** [comment]
> @swapkh91 Can you try 21.11 version of Triton? Also, could you share your model.py so that we can take a closer look?

@Tabrizian this is the `model.py` file. The other 2 models also have similar files. Individually these models are loaded fine in triton, but when I load all 3 I get the `shm-size` error and sometimes 1 out of 3 gets loaded and the other 2 give error. These models are around 80MB in size individually. 

I'll try with 21.11 and update.  

```
import torch
import torchvision
import json
import triton_python_backend_utils as pb_utils
# import torchvision.transforms as transforms
from PIL import Image
import traceback
import datetime

class TritonPythonModel:
    def initialize(self, args):
        try:
            self.model_config = model_config = json.loads(args['model_config'])

            # Get OUTPUT0 configuration
            output0_config = pb_utils.get_output_config_by_name(
                model_config, "OUTPUT__0")

            # Convert Triton types to numpy types
            self.output0_dtype = pb_utils.triton_string_to_numpy(
                output0_config['data_type'])
            
            # Instantiate the PyTorch model
            self.damage_model = torch.jit.load(os.path.dirname(os.path.realpath(__file__))+'/damage.pb')
        except:
            print(str(traceback.format_exc()), flush=True)

    def execute(self, requests):

        output0_dtype = self.output0_dtype

        responses = []

        # Every Python backend must iterate over everyone of the requests
        # and create a pb_utils.InferenceResponse for each of them.
        for request in requests:
            # Get INPUT0
            in_0 = pb_utils.get_input_tensor_by_name(request, "INPUT__0")
            img = in_0.as_numpy()
            img_out = torch.from_numpy(img)

            outs = self.damage_model(img_out)

            out1 = outs[0].cpu().data.numpy()
            out2 = outs[1].cpu().data.numpy()
            out3 = outs[2].cpu().data.numpy()
            out4 = outs[3].cpu().data.numpy()

            # Create output tensors. You need pb_utils.Tensor
            # objects to create pb_utils.InferenceResponse.

            out_tensor_0 = pb_utils.Tensor("OUTPUT__0",
                                           out1.astype(output0_dtype))
            out_tensor_1 = pb_utils.Tensor("OUTPUT__1",
                                           out2.astype(output0_dtype))
            out_tensor_2 = pb_utils.Tensor("OUTPUT__2",
                                           out3.astype(output0_dtype))
            out_tensor_3 = pb_utils.Tensor("OUTPUT__3",
                                           out4.astype(output0_dtype))

            # Create InferenceResponse. You can set an error here in case
            # there was a problem with handling this inference request.
            # Below is an example of how you can set errors in inference
            # response:
            #
            # pb_utils.InferenceResponse(
            #    output_tensors=..., TritonError("An error occured"))
            inference_response = pb_utils.InferenceResponse(
                output_tensors=[out_tensor_0, out_tensor_1, out_tensor_2, out_tensor_3])
            responses.append(inference_response)

        # You should return a list of pb_utils.InferenceResponse. Length
        # of this list must match the length of `requests` list.
        return responses

    def finalize(self):
        """`finalize` is called only once when the model is being unloaded.
        Implementing `finalize` function is optional. This function allows
        the model to perform any necessary clean ups before exit.
        """
        print('Cleaning up...')
```

**Comment by swapkh91** [comment]
@Tabrizian getting the shm error for 21.11

**Comment by Tabrizian** [comment]
Looks like your shared memory size has not been increased and it is 64MBs which is suitable only for loading a single model. I am not an expert in Kubernetes but looks like sizeLimit of empty dir was ignored until recent releases of Kubernetes. https://github.com/kubernetes/kubernetes/issues/63126

@rmccorm4 Do you happen to know how they can set the shared memory size in a Kubernetes Pod?

**Comment by rmccorm4** [comment]
It looks like the memory-backed volume size issue was fixed in k8s v1.20 in this PR: https://github.com/kubernetes/kubernetes/pull/94444. 

See v1.20 changelog as well: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.20.md#feature-6

@swapkh91 Can you try a k8s version >= 1.20 ?

**Comment by swapkh91** [comment]
> It looks like the memory-backed volume size issue was fixed in k8s v1.20 in this PR: [kubernetes/kubernetes#94444](https://github.com/kubernetes/kubernetes/pull/94444).
> 
> See v1.20 changelog as well: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.20.md#feature-6
> 
> @swapkh91 Can you try a k8s version >= 1.20 ?

@rmccorm4 the GKE version is `1.21.5-gke.1302` which means k8s version `1.21`

@Tabrizian 

Also I tried a little workaround

2 out of the 3 models are interdependent (ensemble). So I created a new model as `ensemble` and changed `model.py` to load both and do predict, postprocess on triton itself
```
self.region_model = torch.jit.load('/models/ensemble/1/segmentation.pb')
self.damage_model = torch.jit.load('/models/ensemble/1/damage.pb')
```

Doing this didn't give the `shm error` on docker!

But if I add the 3rd model to the model repo then I'm again getting the error

**Comment by Tabrizian** [comment]
This is expected. I think the shared memory is still not setup properly. Can you share the output of `df -h /dev/shm` when you run the container?

The total shared memory used is equal to number of Python models * 64 MBs. If you have consolidated two of your Python models into a single model, then the shared memory used will be 64 MBs. Alternatively,  you can change the default value to a smaller number if your Python models do not use large tensors. For example, the command below will set it to 64 MBs.

```
/opt/tritonserver/bin/tritonserver --model-repository=`pwd`/models --backend-config=python,shm-default-byte-size=16777216
```

**Comment by swapkh91** [comment]
@Tabrizian when using `ensemble` and `model3` together, i had to set `--shm-size=1g` flag in docker run, below is the output of `df`. Both are completely using the 64M.

```
Filesystem      Size  Used Avail Use% Mounted on
shm             1.0G  128M  896M  13% /dev/shm
```

When using only the `ensemble`

```
Filesystem      Size  Used Avail Use% Mounted on
shm              64M   64M     0 100% /dev/shm
```

I tried the `shm-default-byte-size=16777216` and its working. It works even for 8MB. I'll try passing this argument with kubernetes deployment and update.

One question, would setting a low `shm-default-byte-size` cause any problem in production? Like with greater number of requests to the model.

**Comment by swapkh91** [comment]
@Tabrizian still getting error on kubernetes

Triton is able to load the `ensemble` and `model3` using that flag. But sending request to `model3` after sending 1 request to `ensemble` gives error

GKE logs
```
2021-12-14 13:13:22.908 IST inference failed: [StatusCode.INTERNAL] Failed to process the request(s) for model instance 'model3_0', message: Stub process is not healthy.
2021-12-14 13:13:23.172 IST error reverse proxying request; sockstat: sockets: used 7 TCP: inuse 0 orphan 0 tw 115 alloc 314 mem 68 UDP: inuse 0 mem 2 UDPLITE: inuse 0 RAW: inuse 0 FRAG: inuse 0 memory 0
```
Not sure where the 2nd line is coming from

I set the `--backend-config=python,shm-default-byte-size=16777216`.
Then checked `df` on the container
```
Filesystem      Size  Used Avail Use% Mounted on
shm              64M   32M   32M  50% /dev/shm
```

There is still memory left in `/dev/shm`

**Comment by Tabrizian** [comment]
> One question, would setting a low shm-default-byte-size cause any problem in production? Like with greater number of requests to the model.

If your model has set a value greater than 1 for max_batch_size field and the requests are batched using dynamic batching, then it could run out of shared memory. But if you're input/output size is small I don't think there will be an issue with using smaller value for `shm-default-byte-size`.

> Triton is able to load the ensemble and model3 using that flag. But sending request to model3 after sending 1 request to ensemble gives error

When deploying the model in Kubernetes, do you have multiple replicas?

**Comment by swapkh91** [comment]
@Tabrizian 

> When deploying the model in Kubernetes, do you have multiple replicas?

Just 1 replica of triton. 

```
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  name: "triton-mms-damage"
  annotations:
    cluster-autoscaler.kubernetes.io/safe-to-evict: 'true'
spec:
  predictor:
    minReplicas: 1
    nodeSelector:
      cloud.google.com/gke-nodepool: region-damage-gpu-pool
    triton:
      ports:
        - containerPort: 9000
          name: h2c
          protocol: TCP
      storageUri: <model repo>
      image: <custom image>
      imagePullPolicy: IfNotPresent
      args:
      - --allow-sagemaker=false
      env:
      - name: OMP_NUM_THREADS
        value: "1"
      resources:
        limits:
          nvidia.com/gpu: 1
          memory: "5Gi"
        requests:
          nvidia.com/gpu: 1
          memory: "5Gi"
```

The node (n1, 4cpu, 15G) is also specifically for triton with GPU.

**Comment by Tabrizian** [comment]
@swapkh91 I am not sure what is the underlying issue here. I have filed a ticket for further investigate. Could you please share the content of your model repository and instructions on how to produce this bug? Please include your client file and instructions on how to send requests to the models to produce this bug (e.g. send the request first to model1 then ensemble etc). Thanks.

**Comment by swapkh91** [comment]
@Tabrizian I have emailed you the details

**Comment by Tabrizian** [comment]
Thanks I received the email.

**Comment by akansal1** [comment]
@Tabrizian is there any update on the above issue ?

**Comment by Tabrizian** [comment]
@swapkh91 We've not been able to get to this ticket yet. Are you trying to run a Python model in k8s and faced this error?

**Comment by akansal1** [comment]
@Tabrizian ye we have similar usecase as mentioned in the ticket, trying custom backend for deploying detectron model in production in kserve

**Comment by Tabrizian** [comment]
@akansal1 @swapkh91 Can you try with this fix again and see whether you still face this issue? You need to compile the Python backend from the main branch:

https://github.com/triton-inference-server/python_backend#building-from-source

https://github.com/triton-inference-server/python_backend/commit/e9379464298de8ca9273910fecd07a57202d0789

**Comment by CoderHam** [comment]
@akansal1 @swapkh91 please re-open if the linked fixes do not resolve the issue for you

**Comment by s-rog** [comment]
This is still present in 22.01 when inferencing with a detectron model in python backend
(stub process becomes unhealthy after multiple inference requests)

**Comment by zhyj3038** [comment]
same issue

**Comment by zhyj3038** [comment]
same issue on 21.11 .    
message: Stub process is not healthy.

**Comment by Tabrizian** [comment]
@s-rog Can you open a separate issue and put your model configuration, your model.py, and client file so that we can look into this?

@zhyj3038 Can you try the 22.01 release and share with us whether you are still seeing the same issue?

**Comment by zhyj3038** [comment]
just enlarge the Memory of container in Kubernetes  will solve it

**Comment by s-rog** [comment]
@Tabrizian we managed to get the model working in torchscript (torch backend) and no longer experience this issue

**Comment by swapkh91** [comment]
> @Tabrizian we managed to get the model working in torchscript (torch backend) and no longer experience this issue

@s-rog can you please share how you did it? 
Which triton release?
I guess yours is also a detectron model

**Comment by s-rog** [comment]
I was not the one on my team who implemented it but I think [this comment in #2024](https://github.com/triton-inference-server/server/issues/2024#issuecomment-974058244) is the solution. We had it working on 21.09

**Comment by Tabrizian** [comment]
@s-rog / @swapkh91 if you could share you're still facing the same issue in 22.01 with Python backend please share your model so that we can fix this. Thank you.

**Comment by swapkh91** [comment]
> @s-rog / @swapkh91 if you could share you're still facing the same issue in 22.01 with Python backend please share your model so that we can fix this. Thank you.

@Tabrizian getting same error in 22.01 when using python backend. The details to replicate are in the email I sent earlier, change the triton version when creating custom triton image

```
Traceback (most recent call last):
  File "test_triton.py", line 26, in <module>
    out2 = triton_client.infer(model2, inputs,request_id='1',model_version='1')
  File "/home/swapnesh/triton_env/lib/python3.8/site-packages/tritonclient/grpc/__init__.py", line 1146, in infer
    raise_error_grpc(rpc_error)
  File "/home/swapnesh/triton_env/lib/python3.8/site-packages/tritonclient/grpc/__init__.py", line 62, in raise_error_grpc
    raise get_error_grpc(rpc_error) from None
tritonclient.utils.InferenceServerException: [StatusCode.INTERNAL] Failed to process the request(s) for model instance 'maskrcnn_0', message: Stub process is not healthy.
```

![Screenshot from 2022-03-03 14-50-10](https://user-images.githubusercontent.com/5755405/156534911-0d833ab1-e8e8-4456-8655-359c0641caad.png)


Node config
![Screenshot from 2022-03-03 14-53-39](https://user-images.githubusercontent.com/5755405/156535374-7944b599-3e6d-43f2-9fc9-df78901ae3c0.png)

**Comment by farzanehnakhaee70** [comment]
Hi @swapkh91 
Is your problem solved?

**Comment by swapkh91** [comment]
> Hi @swapkh91 Is your problem solved?

@farzanehnakhaee70 I have'nt tried with latest triton version. But it wasn't working with 22.01

**Comment by farzanehnakhaee70** [comment]
Thanks @swapkh91 
This is a problem currently blocks me. I would appreciate if you can find any solution for that @Tabrizian .

**Comment by JJwangbilin** [comment]
@farzanehnakhaee70 @swapkh91 @Tabrizian   
I had the same problem，Stub process is unhealthy and it will be restarted 
env:
triton:22.02-py3
k8s:1.19

but, docker run image --shm-size=128Mi,  its running work。only run on k8s has this problem，

**Comment by EgShes** [comment]
same problem on 22.04 while deploying on kubernetes
```
12:46:27.395847 1 model_repository_manager.cc:1355] failed to load 'ischemia' version 1: Internal: Unable to initialize shared memory key 'triton_python_backend_shm_region_4' to requested size (67108864 bytes). If you are running Triton inside docker, use '--shm-size' flag to control the shared memory region size. Each Python backend model instance requires at least 64MBs of shared memory. Error: No such file or directory
```

Pod is allowed to take memory up to 20Gb

**Comment by BinhMinhs10** [comment]
I added ipc config and it worked

**Comment by foocker** [comment]
> I added ipc config to docker and it worked
how?

**Comment by rl02898** [comment]
> > I added ipc config to docker and it worked
> > how?

I was able to add 'ipc: host' to my docker compose service and it runs. It isn't ideal, but this is the explanation from Docker's documentation:

> IPC (POSIX/SysV IPC) namespace provides separation of named shared memory segments, semaphores and message queues.

> Shared memory segments are used to accelerate inter-process communication at memory speed, rather than through pipes or through the network stack. Shared memory is commonly used by databases and custom-built (typically C/OpenMPI, C++/using boost libraries) high performance applications for scientific computing and financial services industries. If these types of applications are broken into multiple containers, you might need to share the IPC mechanisms of the containers, using "shareable" mode for the main (i.e. “donor”) container, and "container:<donor-name-or-ID>" for other containers.

https://docs.docker.com/engine/reference/run/

**Comment by lucidyan** [comment]
You should definitely keep track of the memory **inside the container** using `top` or `htop` while loading triton, because in my case it was a hint. I found that I didn't have enough memory for a specific model inside the container, and increased it in [docker settings](https://stackoverflow.com/questions/44533319/how-to-assign-more-memory-to-docker-container). After that problem was solved.

Solutions from the thread with `shm-size=Xgb` and `ipc=host` docker's flags didn't work for me, because in my case it had nothing to do with shared memory. Also worth mentioned, that triton explicitly complains about lack of shared memory in that case.
